New winrm upload functionality

Lantero opened the pull request:
Two threads here:

- New `winrm_upload` function:

  This allows file transfer through WinRM sreaming. It's not speedy for big binary files, but works well for 
  text/config files. There are unit and integration tests covering it.

  The origin of this PR is that we need to provide sysprep unattend.xml files to be able to change the SID 
  in windows machines in a fully automated way. It's a bit frustrating, but sysprep is the only supported 
  tool that performs that task, and the only way to fully automate it is to provide config in xml form that 
  gets read after the reboot (Otherwise you get undesired side effects like Administrator password reset 
  that basically breaks the downstream of your automation pipeline). The reason behind changing SID's 
  is that they have to be unique if you want to have them in the same domain, and cloning windows 
  boxes from a Vsphere template sadly gives them the same SID. It all looks a bit too complicated, but 
  this opens a big door for further Windows automation.

- Fix and improve integration tests:
  I have fixed some race conditions that appeared recently in the integration tests. I've also made 
  them faster by optimizing the number of times we create and delete the vms and I've made Jenkins to 
  run them in parallel (Python 2 and 3). Builds are now ~5 times faster: 14 minutes against 1 hour and 6 
  minutes.

Lantero committed:
Remove deprecated CHANGELOG from MANIFEST.in

Lantero committed:
Add winrm_upload function

Lantero committed:
Add integration test for the winrm upload function

Lantero committed:
Update README.md

Lantero committed:
Add unit test and make the function un only when there is a vm object

Lantero committed:
Bump version in setup.py

Lantero committed:
Fix unit test for python 3

Lantero committed:
Improve winrm_upload function and fix tests

Lantero committed:
Fix flake8

Lantero committed:
Fix problem with vm name

Lantero committed:
Fix integration tests

Lantero committed:
Run tests in parallel in the Jenkinsfile

Lantero committed:
Fix quotes in Jenkinsfile

Lantero committed:
Add missing config wrapper in the Jenkinsfile

Lantero committed:
Add missing folder env var

Lantero committed:
Increase shutdown timeout in tests

Lantero committed:
Skip test that only passes if you are a user with high privileges

Lantero committed:
Use cd to avoid conflicts with the filesystem in the parallel integration tests

Lantero committed:
Fix typo: It's dir, not cd

Lantero committed:
Fix paths in the Jenkinsfile

Lantero committed:
Use separate workspaces also for unit tests

Lantero committed:
Fix all broken paths

edsharp reviewed (commented):
Carlos, poor you, it's hideous!! :$

I imagine you searched long and hard to find an alternative. From what I can see, `Copy-Item` only works between PowerShell sessions which we don't have here.

Are there file size limitations you know of?

I'll give this a try. It may be that this is what we have to do, but I wish it weren't!

edsharp reviewed (changes requested):
```vm.winrm_upload('C:\\jre-8u144-windows-x64-3.exe', '/Users/edward/Downloads/jre-8u144-windows-x64.exe')
Copying "/Users/edward/Downloads/jre-8u144-windows-x64.exe" to "C:\jre-8u144-windows-x64-3.Copying "/Users/edward/Downloads/jre-8u144-windows-x64.exe" to "C:\jre-8u144-windows-x64-3.Copying "/Users/edward/Downloads/jre-8u144-windows-x64.exe" to "C:\jre-8u144-windows-x64-3.Copying "/Users/edward/Downloads/jre-8u144-windows-x64.exe" to "C:\jre-8u144-windows-x64-3.exe" ... [                              ]   0 %Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "vcdriver/config.py", line 114, in wrapper
    return function(*args, **kwargs)
  File "vcdriver/vm.py", line 395, in winrm_upload
    raise WinRmError(script, status_code)
vcdriver.exceptions.WinRmError: Remote execution of "add-content -value $([System.Convert]::FromBase64String("FRtlAgBEi8hMiwUJZQIASIvTSI2NmAEAAOgabwAAkEiNDbqkAgBIiY1IAQAASI0NdKQCAEiJjVABAADHhVgBAAAiAAAATI2FSAEAAEiL0EiNjcABAADooSYAAEiNFSYyAwBIjY3AAQAA6OIxAQCQSItMJFhIhcl0BejCAgEASIvDSIuNAAIAAEgzzOjg+AAATI2cJBADAABJi1soSYtzMEmL40FcX13DzMzMzOlL/QAAzMzMSIlMJAhXSIPsMEjHRCQg/v///0iJXCRISIvaSIv56HH9AACQSI0FGagCAEiJB0iNTxhIx0EYDwAAAEiDYRAAxgEASI1TGEmDyf9FM8DoohcAAJBIi8dIi1wkSEiDxDBfw8zMzEBTSIPsIEiL2egi/QAATI0ds6cCAEyJG0iLw0iDxCBbw8zMzEiJXCQIV0iD7CBIjQWrpwIAi/pIi9lIiQFIg3kwEHIJSItJGOjbAQEASINjKABIx0MwDwAAAEiLy8ZDGADofvwAAED2xwF0CEiLy+i0AQEASIvDSItcJDBIg8QgX8PMzEiLxFVWV0iNqDj///9IgeywAQAASMdEJGD+////SIlYGEiLBaRKAwBIM8RIiYWgAAAASIv6SIvZSIlMJFgz9ol0JCBIOTVqawMAdUZIjQXxogIASIlEJChIjQWdowIASIlEJDDHRCQ4KwAAAEyNRCQoSI0VrKMCAEiNTWDoi/n//0iNFcwwAwBIjU1g6CMwAQDMSI1MJHDoVAsAAJBIjUQkKEiJRCRQSMdEJEAPAAAASIl0JDhAiHQkKEmDyf9FM8BIi9dIjUwkKOg+FgAAkEiNFV6jAgBIjUwkcOgEJQAAkEiNVCQoSI1MJHDoACQAAEiNFT2jAgBIi8jo5SQAAL4BAAAAOTWmagMAfkiNfgdIjRUiowIASI1MJHDoxCQAAEyLHY1qAwBKixQfSI1MJHDoryQAAEiNFfiiAgBIjUwkcOieJAAA/8ZIg8cIOzVeagMAfLtIi9NIjUwkcOhz9///x0QkIAEAAABIjUwkcOh99///SIvDSIuNoAAAAEgzzOhP9gAASIucJOABAABIgcSwAQAAX15dw8zMzMxIiUwkCFdIg+wwSMdEJCD+////SIlcJEhIi9pIi/no7foAAJBIjQVlpQIASIkHSI1PGEjHQRgPAAAASINhEADGAQBIjVMYSYPJ/0UzwOgeFQAAkEiLx0iLXCRISIPEMF/DzMzMSIlcJAhXSIPsIEiNBRulAgCL+kiL2UiJAUiDeTAQcglIi0kY6Hv/AABIg2MoAEjHQzAPAAAASIvLxkMYAOge+gAAQPbHAXQISIvL6FT/AABIi8NIi1wkMEiDxCBfw8zMSIvEVQ==")) -encoding byte -path C:\jre-8u144-windows-x64-3.exe" failed with exit code 1```

edsharp commented:
http://download.oracle.com/otn-pub/java/jdk/8u144-b01/090f390dda5b47b9b721c7dfaa008135/jre-8u144-windows-x64.exe

I ran a few tests of different files. When I started to upload the `jre-8u144-windows-x64.exe`, it failed with an error similar to above. It failed in different places, sometimes after 4K, after 20K or in its furthest run after 208K.

(As an aside it copied at slightly over 2KB/s)

edsharp commented on vcdriver/vm.py:
```diff
@@ -342,6 +346,73 @@ def winrm(self, script, winrm_kwargs=dict(), **kwargs):
+                **winrm_kwargs
+            )
+            with open(local_path, 'rb') as f:
+                contents = f.read()
```
Why read the whole file in advance rather than reading it in chunks?

Lantero replied:
That's a good point, I can fix that. I guess I wanted to get the whole size before for the progress bar, but that can be done by reading the metadata with `os.stats` instead.

Lantero commented:
Alternatives are:
- Install SSH, which is a bit clunky in Windows, as you need to do it through 3rd parties
- Setup an SFTP server with Windows IIS
- Mount a file share

All of them require adding extra stuff to our windows templates or doing it at runtime, which seems to be overkill to send a config file, which is the use case here. So I thought re-using WinRM could do the trick, as discussed here https://github.com/diyan/pywinrm/issues/18. I knew it was slow, but I was not expecting it to fail with binaries (The integration test creates a 5kb binary file of zeroes and checks the SHA256 and seems to be working :disappointed: ).

I'll take a look to see if I can improve it.

Lantero commented:
By the way, speed is limited by the maximum size that WinRM allows for remote powershell scripts, which is something of the same magnitude of the `steps` default value of the `winrm_upload` function. So I don't think we can do much better there unless I am missing something.

Lantero committed:
Read file by chunks

Lantero commented:
I cannot find a reliable way of guessing whether a file is a text file or not. Some guys recommend using mimetypes, some others looking at null bytes. I'll give a test to that java exe to see if I can debug the error.

Lantero committed:
Run cron job only for master

Lantero committed:
Add coverage reports

Lantero committed:
Add useful pipeline options

Lantero committed:
Get better debug information on exceptions

Lantero committed:
Add cleanup stage

Lantero committed:
Abstract out common winrm functionality

Lantero committed:
Fix problem when windows process keeps lock on file + test

Lantero committed:
Fix typo in html coverage report path

Lantero commented:
@edsharp I have found the problem and I think I have fixed it. Apparently there's a race condition where you can have the previous write process still open when you try to add more bytes to the file and that makes it fail, so you need to provide a small delay for that case. I tested it with your java binary until I got tired of waiting after 10-15 minutes with no errors at 2% and 1MB (It would have taken hours at such speed), and then I tried other binaries and they all went good, with the right sha256.

I have added after your last review:
- File being read in chunks
- Fix for the race condition and the test to cover that case
- Better debug information for the `RemoteCommandError` exceptions so they also show stdout/stderr and not just the command that failed and the status code
- Some other fixes/improvements to the Jenkinsfile that I saw on the fly

edsharp commented:
Hmm first attempt failed: 
```
vm.winrm_upload(local_path='/Users/edward/Downloads/jre-8u144-windows-x64.exe',remote_path='c:\\jre-8u144-windows-x64.exe')
vcdriver_vm_winrm_username: Administrator
vcdriver_vm_winrm_password:
Copying "/Users/edward/Downloads/jre-8u144-windows-x64.exe" to "c:\jre-8u144-windows-x64.exe" ... [                              ]   0 %Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "vcdriver/config.py", line 114, in wrapper
    return function(*args, **kwargs)
  File "vcdriver/vm.py", line 398, in winrm_upload
    raise WinRmError(script, status_code)
vcdriver.exceptions.WinRmError: Remote execution of "add-content -value $([System.Convert]::FromBase64String("aOgaEAAAi0MoiYQkkAAAAEiNVCRoSI1MJDjo4YsAAJBIg7wkgAAAABByCkiLTCRo6Av7AABIibQkgAAAAEiJfCR4QIh8JGhIg7wkoAEAABByDUiLjCSIAQAA6OH6AABMjYQkYAEAAEiNlCTAAAAASI2MJJgAAADorGsAAJAzyf8Vq1wCAEiJtCQIAQAASIm8JAABAABAiLwk8AAAAEiJtCQ4AQAASIm8JDABAABAiLwkIAEAAEyLwLpkAAAASI2MJPAAAADo8qwAAJBIjZQkmAAAAEiNjCTwAAAA6JyxAACQSIO8JDgBAAAQcg1Ii4wkIAEAAOhH+gAASIm0JDgBAABIibwkMAEAAECIvCQgAQAASIO8JAgBAAAQcg1Ii4wk8AAAAOgX+gAASI2UJMAAAABIjUwkaOjRhgAAkEiLVCRwSItMJGhIK9FIuGdmZmZmZmZmSPfqSMH6BEiLwkjB6D9IA9C7AQAAAEg703URSI2UJJgAAADorcoAAITAdQNAit9IjUwkaOgEDQAAhNt0QEiNlCSYAAAASI1MJGjo+vf//5BIjVQkaEiNjCSYAAAA6Ef6//+JRCQwSIO8JIAAAAAQcklIi0wkaOhy+QAA6z26AwAAADkVvV4DAH8wSI0FfJwCAEiJRCQoSI0FoJwCAEiJRCQgRI1Kd0yNBeCaAgBIjQ2RXgMA6Izy//+QSIO8JLAAAAAQcg1Ii4wkmAAAAOgb+QAASIm0JLAAAABIibwkqAAAAECIvCSYAAAASI1MJDjo/YYAAJBIjUwkOOg+CgAAkEiDvCTYAAAAEHINSIuMJMAAAADo1fgAAEiJtCTYAAAASIm8JNAAAABAiLwkwAAAAEiDvCR4AQAAEHIOSIuMJGABAADopfgAAJCLRCQwSIuMJLgBAABIM8zowO4AAEyNnCTAAQAASYtbEEmLcxhJi+Nfw8zMzEiD7DhIjQ2KmQIA/xUvWgIASINkJCAATI1EJCBIjRV1mQIAuQIAAAD/FfpZAgCFwHQeSItMJCBIjRVqmQIA/xUcWgIASIXAdAe5AAgAAP/QSIPEOOkH/P//zMzMQFNIg+wgSIvZSIsJSIXJdAXo+vYAAEiDIwBIg8QgW8NIi0H4SGNQBEiNBUWcAgBIiUQK+MPMzMxIjQWVmwIASIkB6QniAADMSIlcJAhXSIPsIEiLeWhIjQWXmwIASIvZSIkBSIX/dBBIi8/oxOIAAEiLz+ik9wAASI1LCEiLXCQwSIPEIF/pgegAAMxIg8EI6ZfoAADMzMxIg8EI6ZPoAADMzMwzwMPMg8j/w0BTSIPsIEiLAUiL2f9QMIPK/zvCdBdIi0NYARBIi0tASIsRSI1CAUiJAQ+2EovCSA==")) -encoding byte -path c:\jre-8u144-windows-x64.exe" failed with exit code 1
```
